### PR TITLE
grid: cleaning up potential zombie streams

### PR DESCRIPTION
**Pull request by alyssawilk**
Fixing an issue where "prefetched" H3 streams were not cleaned up promptly.

Risk Level: low
Testing: new e2e tests
Docs Changes: n/a
Release Notes: inline
[Optional Runtime guard:] yes

### CODE REVIEW

**Comment by alyssawilk** [comment]
this may still have a gcc test flake but I think it's ready for a review pass

File: test/extensions/filters/http/alternate_protocols_cache/filter_integration_test.cc [hunk 1]
```diff
@@ -144,6 +160,125 @@ TEST_P(FilterIntegrationTest, AltSvc) {
...
+  const uint64_t request_size = 0;
+  const uint64_t response_size = 0;
+  const std::chrono::milliseconds timeout = TestUtility::DefaultTimeout;
+
+  initialize();
+  codec_client_ = makeHttpConnection(makeClientConnection((lookupPort("http"))));
+
+  absl::Notification block_until_notify;
+  // Block the H3 upstream so it can't process packets.
+  fake_upstreams_[1]->runOnDispatcherThread([&] { block_until_notify.WaitForNotification(); });
```
Comment (RyanTheOptimist) [inline test/extensions/filters/http/alternate_protocols_cache/filter_integration_test.cc:202]: This is such a great/clever hack! I love it.

File: test/extensions/filters/http/alternate_protocols_cache/filter_integration_test.cc [hunk 2]
```diff
@@ -144,6 +160,125 @@ TEST_P(FilterIntegrationTest, AltSvc) {
...
+TEST_P(FilterIntegrationTest, AltSvcCachedH3Slow) {
+  // Start with the alt-svc header in the cache.
+  fill_cache_ = true;
+
+  const uint64_t request_size = 0;
+  const uint64_t response_size = 0;
+  const std::chrono::milliseconds timeout = TestUtility::DefaultTimeout;
+
+  initialize();
+  codec_client_ = makeHttpConnection(makeClientConnection((lookupPort("http"))));
```
Comment (RyanTheOptimist) [inline test/extensions/filters/http/alternate_protocols_cache/filter_integration_test.cc:198]: nit: is there an extra set of ()s around `lookupPort`? Here and elsewhere.

**Review by RyanTheOptimist — COMMENTED** [review]
Looks great! Thanks for doing this. A few small comments.

File: test/extensions/filters/http/alternate_protocols_cache/filter_integration_test.cc [hunk 3]
```diff
@@ -67,6 +67,18 @@ name: alternate_protocols_cache
     HttpProtocolIntegrationTest::initialize();
   }
 
+  void writeFile(uint32_t index) {
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply test/extensions/filters/http/alternate_protocols_cache/filter_integration_test.cc]: done

File: test/extensions/filters/http/alternate_protocols_cache/filter_integration_test.cc [hunk 4]
```diff
@@ -105,6 +120,7 @@ name: alternate_protocols_cache
     }
     throw EnvoyException("Failed to find a port after 10 tries");
   }
+  bool fill_cache_ = false;
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply test/extensions/filters/http/alternate_protocols_cache/filter_integration_test.cc]: done

File: test/extensions/filters/http/alternate_protocols_cache/filter_integration_test.cc (hunk 2 above)
Comment (alyssawilk) [reply test/extensions/filters/http/alternate_protocols_cache/filter_integration_test.cc]: yep!

**Review by RyanTheOptimist — DISMISSED** [review]

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**